include/v4l2_camera/rate_bound_status.hpp

```diff
@@ -0,0 +1,281 @@
+    }
+
+    // Update the current state if
+    // - immediate error report is required and the observed state is error
+    // - Or the same state is observed multiple times
+    if ((immediate_error_report_ && std::holds_alternative<Error>(candidate_state_)) ||
+        (get_num_observation(candidate_state_) >= num_frame_transition_)) {
+      current_state_ = candidate_state_;
+      std::visit([](auto& s) {
+        s.num_observation = 1;
+      }, candidate_state_);
+    }
+
+    stat.summary(get_level(current_state_), get_msg(current_state_));
+
+    std::stringstream ss;
+    ss << std::fixed << std::setprecision(2) << frequency_.value_or(0.0);
+    stat.add("Publish rate", ss.str());
+
+    ss.str("");  // reset contents
+    ss << get_level_string(get_level(frame_result));
+    stat.add("Rate status", ss.str());
+
+    ss.str("");  // reset contents
+    ss << std::fixed << std::setprecision(2) << ok_params_.min_frequency;
+    stat.add("Minimum OK rate threshold", ss.str());
+
+    ss.str("");  // reset contents
+    ss << std::fixed << std::setprecision(2) << ok_params_.max_frequency;
+    stat.add("Maximum OK rate threshold", ss.str());
+
+    ss.str("");  // reset contents
+    ss << std::fixed << std::setprecision(2) << warn_params_.min_frequency;
+    stat.add("Minimum WARN rate threshold", ss.str());
+
+    ss.str("");  // reset contents
+    ss << std::fixed << std::setprecision(2) << warn_params_.max_frequency;
+    stat.add("Maximum WARN rate threshold", ss.str());
+
+    ss.str("");  // reset contents
+    ss << get_num_observation(candidate_state_);
+    stat.add("Observed frames", ss.str());
+
+    ss.str("");  // reset contents
+    ss << num_frame_transition_;
+    stat.add("Observed frames transition threshold", ss.str());
+  }
+
+protected:
+  RateBoundStatusParam ok_params_;
+  RateBoundStatusParam warn_params_;
+  size_t num_frame_transition_;
+  bool immediate_error_report_;
+  bool zero_seen_;
+  std::optional<double> frequency_;
+  std::optional<double> previous_frame_timestamp_;
+  std::mutex lock_;
+
+  StateHolder candidate_state_;
+  StateHolder current_state_;
+
+  unsigned char get_level(const StateHolder& state) {
+    return std::visit([](const auto& s){return s.level;}, state);
+  }
+
+  size_t get_num_observation(const StateHolder& state) {
+    return std::visit([](const auto& s){return s.num_observation;}, state);
+  }
+
+  std::string get_msg(const StateHolder& state) {
+    return std::visit([](const auto& s){return s.msg;}, state);
+  }
+
+  std::string get_level_string(unsigned char level) {
+    switch(level) {
+      case diagnostic_msgs::msg::DiagnosticStatus::OK:
+        return "OK";
+      case diagnostic_msgs::msg::DiagnosticStatus::WARN:
+        return "WARN";
+      case diagnostic_msgs::msg::DiagnosticStatus::ERROR:
+        return "ERROR";
+      case diagnostic_msgs::msg::DiagnosticStatus::STALE:
+        return "STALE";
+      default:
+        return "UNDEFINED";
+    }
+  }
```

Comment:
These can all be made `static`

Reply:
Thanks for catching this. I changed these to static in 70be4db